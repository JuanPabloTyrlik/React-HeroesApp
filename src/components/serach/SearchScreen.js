import React, { useState } from 'react';
import { HeroCard } from '../heroes/HeroCard';
import { useHistory, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { getHeroByName } from '../../selectors/getHeroByName';
import { useMemo } from 'react';

export const SearchScreen = () => {
    const history = useHistory();
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    const [searchValue, setSearchValue] = useState(q);

    const heroesFiltered = useMemo(() => getHeroByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchValue}`);
    };

    const handleChange = (e) => {
        setSearchValue(e.target.value);
    };

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Find your Hero"
                            className="form-control"
                            style={{ marginBottom: '1rem' }}
                            autoComplete="off"
                            onChange={handleChange}
                            value={searchValue}
                        />
                        <button
                            type="submit"
                            className="btn btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    <div className="animate__animated animate__fadeIn">
                        {q ? (
                            heroesFiltered.length === 0 && (
                                <div className="alert alert-warning">
                                    No results for {q}
                                </div>
                            )
                        ) : (
                            <div className="alert alert-info">
                                {' '}
                                Search a Hero{' '}
                            </div>
                        )}
                        {heroesFiltered.map((hero) => (
                            <div className="mb-3" key={hero.id}>
                                <HeroCard key={hero.id} {...hero} />
                            </div>
                        ))}
                    </div>
                </div>
            </div>
        </div>
    );
};
