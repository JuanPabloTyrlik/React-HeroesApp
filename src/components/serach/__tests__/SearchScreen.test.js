import { mount } from 'enzyme';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Route, Router } from 'react-router-dom';
import { SearchScreen } from '../SearchScreen';

describe('Tests on SearchScreen', () => {
    const history = createMemoryHistory();
    test('should render correctly', () => {
        history.push('/search');
        const wrapper = mount(
            <Router history={history}>
                <Route exact path="/search" component={SearchScreen} />
            </Router>
        );
        expect(wrapper).toMatchSnapshot();
    });
    test('should search Batman and return results using queryString', () => {
        history.push('/search?q=Batman');
        const wrapper = mount(
            <Router history={history}>
                <Route exact path="/search" component={SearchScreen} />
            </Router>
        );
        expect(wrapper.find('input').prop('value')).toMatch('Batman');
        expect(wrapper.find('img').prop('alt')).toMatch('Batman');
    });
    test('should search a hero that does not exist and show a message', () => {
        history.push('/search?q=NotAHero');
        const wrapper = mount(
            <Router history={history}>
                <Route exact path="/search" component={SearchScreen} />
            </Router>
        );
        expect(wrapper.find('input').prop('value')).toMatch('NotAHero');
        expect(wrapper.find('div.alert.alert-warning').text()).toMatch(
            'No results for NotAHero'
        );
    });
    test('should search a hero and update the queryString', () => {
        history.push('/search');
        const wrapper = mount(
            <Router history={history}>
                <Route exact path="/search" component={SearchScreen} />
            </Router>
        );
        history.push = jest.fn();
        wrapper
            .find('input')
            .simulate('change', { target: { value: 'Punisher' } });
        wrapper.find('form').simulate('submit', { preventDefault() {} });
        expect(history.push).toHaveBeenCalledWith('?q=Punisher');
    });
});
