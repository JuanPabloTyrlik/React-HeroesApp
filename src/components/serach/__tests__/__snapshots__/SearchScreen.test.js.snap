// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on SearchScreen should render correctly 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          onChange={[Function]}
          placeholder="Find your Hero"
          style={
            Object {
              "marginBottom": "1rem",
            }
          }
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeIn"
      >
        <div
          className="alert alert-info"
        >
           
          Search a Hero
           
        </div>
      </div>
    </div>
  </div>
</div>
`;
